File: .github/atoma/tools/scripts/resolve_notify.py
#!/usr/bin/env python3
"""Resolve who to notify (a GitHub login) for a given issue.

Looks for an `<!-- atoma:notify=LOGIN -->` tag in the issue body -- embedded
by atoma_github_mcp_server.py when the agent that created the issue knew who
the original human requester was (see ISSUE_NOTIFY / _notify_tag_prefix).

Falls back to the issue's own author when no tag is present and the author
is a human. This covers root issues, which are opened directly by a human
and therefore never carry the tag themselves (only bot-created sub-issues
and PRs do).

Usage:
    resolve_notify.py --repo OWNER/REPO --number N
Prints the resolved login (possibly empty) to stdout. Never raises for
missing data -- callers treat an empty result as "nobody to notify".
"""
import argparse
import json
import re
import subprocess

NOTIFY_RE = re.compile(r"<!--\s*atoma:notify=([A-Za-z0-9-]+)\s*-->")


def gh_json(*args):
    try:
        out = subprocess.run(
            ["gh", *args], capture_output=True, text=True, check=True
        ).stdout
        return json.loads(out) if out.strip() else {}
    except (subprocess.CalledProcessError, json.JSONDecodeError):
        return {}


def main():
    p = argparse.ArgumentParser()
    p.add_argument("--repo", required=True)
    p.add_argument("--number", required=True, type=int)
    args = p.parse_args()

    d = gh_json(
        "issue", "view", str(args.number), "--repo", args.repo,
        "--json", "body,author",
    )
    body = d.get("body") or ""
    match = NOTIFY_RE.search(body)
    if match:
        print(match.group(1))
        return

    author = d.get("author") or {}
    if (author.get("type") or "").upper() == "USER":
        login = author.get("login") or ""
        if login:
            print(login)
            return

    print("")


if __name__ == "__main__":
    main()

File: .github/atoma/tools/scripts/resolve_notify.ts
#!/usr/bin/env bun
/**
 * resolve_notify.ts — Resolve who to notify (a GitHub login) for a given
 * issue or PR.
 *
 * Looks for an `<!-- atoma:notify=LOGIN -->` tag in the body -- embedded by
 * mcp/github.ts when the agent that created the issue/PR knew
 * who the original human requester was.
 *
 * Falls back to the issue/PR's own author when no tag is present and the
 * author is a human. If neither is available, walks up the
 * `atoma:parent=#N` / `atoma:parent-issue=N` chain and retries on the
 * parent, since every sub-issue/PR is ultimately rooted in an issue a human
 * opened directly. Gives up after MAX_HOPS to guard against cycles.
 *
 * Usage:
 *   resolve_notify.ts --repo OWNER/REPO --number N
 *
 * Prints the resolved login (possibly empty) to stdout. Never throws for
 * missing data -- callers treat an empty result as "nobody to notify".
 */
import { parseArgs } from "node:util";
import { gh } from "./lib/gh.ts";

/** CLI contract for this script, used by callers (e.g. src/workflows/*.wac.ts) to build a type-checked argv. */
export interface ResolveNotifyArgs {
  repo: string;
  number: string | number;
}

const NOTIFY_RE = /<!--\s*atoma:notify=([A-Za-z0-9-]+)\s*-->/;
const PARENT_RE = /<!--\s*atoma:parent(?:-issue)?=#?(\d+)\s*-->/;
const MAX_HOPS = 10;

interface IssueLookup {
  body?: string;
  login?: string;
  type?: string;
}

function fetch(repo: string, number: number): IssueLookup {
  const { code, stdout } = gh(
    "api", `repos/${repo}/issues/${number}`,
    "--jq", "{body: .body, login: .user.login, type: .user.type}",
  );
  if (code !== 0 || !stdout.trim()) return {};
  try {
    return JSON.parse(stdout) as IssueLookup;
  } catch {
    return {};
  }
}

function resolve(repo: string, number: number): string {
  const visited = new Set<number>();
  let current = number;
  for (let i = 0; i < MAX_HOPS; i++) {
    if (visited.has(current)) break; // cycle guard
    visited.add(current);

    const d = fetch(repo, current);
    const body = d.body ?? "";

    const match = NOTIFY_RE.exec(body);
    if (match) return match[1]!;

    if ((d.type ?? "").toLowerCase() === "user" && d.login) {
      return d.login;
    }

    const parentMatch = PARENT_RE.exec(body);
    if (!parentMatch) break;
    current = Number(parentMatch[1]);
  }
  return "";
}

function main(): void {
  const { values } = parseArgs({
    args: Bun.argv.slice(2),
    options: {
      repo: { type: "string" },
      number: { type: "string" },
    },
  });

  if (!values.repo || !values.number) {
    console.error("usage: resolve_notify.ts --repo OWNER/REPO --number N");
    process.exit(2);
  }

  console.log(resolve(values.repo, Number(values.number)));
}

if (import.meta.main) main();
